### DISCUSSION TITLE
Russian model proposal

### DISCUSSION THREAD

**Discussion by kuk** [explosion/spaCy#6628] [Language Support]
cc @buriy

We trained high quality Russian model for SpaCy and would like to propose it for official SpaCy models registry.

- General description, links to evaluation tables https://github.com/natasha/natasha-spacy#natasha-spacy
- Links to training data, commands to reproduce our results https://github.com/natasha/natasha-spacy#training
- Trained model, usage example https://github.com/natasha/natasha-spacy#usage

All training sources are available under MIT license. We adapt datasets and embeddings table to fit SpaCy utilities. Training procedure uses standart `spacy convert`, `spacy init-model`, `spacy train`.

Could you please take a look at https://github.com/natasha/natasha-spacy? Is it possible to add such model to official SpaCy models registry? Is there anything we can do help you with this process? For example, we could contribute tests to spacy-models repo or main spacy repo.

**Comment by buriy** [comment]
@honnibal @adrianeboyd 
Could you please take a look at this and suggest our further steps.

**Comment by honnibal** [comment]
This looks really cool! The results are certainly impressive. I do think we'll be able to integrate it as an official model, especially since it's MIT licensed.

The main requirement we have for distributing the models is that we need to add them to our automation systems, so that we can train the model ourselves. It looks like it'll be pretty easy to do that, the main annoyance being how bad the json format is in v2, which makes the training corpus hard to work with.

For spaCy v3, the model training process is now much simpler: we use the `spacy project` command to manage the various bits of preprocessing required, and use a config file to encapsulate all the settings. This will allow us to work from the source data as well, which is cool. We'll also be able to finally open-source the model training stuff, so that people can make direct pull-requests to add models.

The thing that would be most helpful from your side is to make sure it still works with spaCy v3, using the current `spacy-nightly`. The conversion now uses the `DocBin` binary format, so you might even be able to train on more text, without hitting the conversion problems.

Quick question about your evaluation: is the evaluation also on silver-standard data, or is the development and test data gold-standard?

**Comment by kuk** [comment]
> The thing that would be most helpful from your side is to make sure it still works with spaCy v3

Sure thing, I will check it. How to reproduce `spacy train --base-model ...` from v2? We use pruned embeddings for initialization `spacy init vectors ru data/navec.12B.300d.txt.gz train/vectors --prune 125000`, then I do `spacy train config.cfg --paths.vectors train/vectors`, but final model uses embeddings trained from scratch, ignoring `train/vectors`.

> The main requirement we have for distributing the models is that we need to add them to our automation systems, so that we can train the model ourselves. It looks like it'll be pretty easy to do that, the main annoyance being how bad the json format is in v2, which makes the training corpus hard to work with.

We use only standard SpaCy commands `spacy convert`, `spacy init-model`, `spacy train` so that it is easier for you to integrate our work into your pipeline. I can think of 2 things that might be problematic, both due to training corpus size:
1.  `spacy convert` requires 32 GB of RAM 
2. `spacy train` take 2 hours per iteration on CPU

> Quick question about your evaluation: is the evaluation also on silver-standard data, or is the development and test data gold-standard?

Yep, evaluation is also silver-standard. But we also check the final model on a number of Russian gold-standart datasets, see <a href="https://github.com/natasha/naeval#morphology-taggers">morphology</a>, <a href="https://github.com/natasha/naeval#syntax-parser">syntax</a>, and <a href="https://github.com/natasha/naeval#ner">NER</a> sections of our repo dedicated to public models evaluations.

**Comment by kuk** [comment]
> The thing that would be most helpful from your side is to make sure it still works with spaCy v3

Added SpaCy v3 support using SpaCy Projects.

- config.cfg + project.yml https://github.com/natasha/natasha-spacy/tree/master/project
- v3 model + usage example https://github.com/natasha/natasha-spacy/blob/master/README.md#usage
- `project run` commands https://github.com/natasha/natasha-spacy/blob/master/README.md#v3
- package https://github.com/natasha/natasha-spacy/blob/master/README.md#v3-1

> How to reproduce spacy train --base-model ... from v2?

I think the problem was in config. By default `initialize.vectors` is set to null, changed it to `vectors = ${paths.vectors}`.

**Reply by honnibal** [reply]
Awesome, thanks! We shouldn't need anything else from you I think -- this should let us support Russian really nicely!

**Reply by kuk** [reply]
1. Could you please share your plans, when can we expect Russian model support? There is no hurry, just curious.
2. Do you plan to train Russian model both for v2 and v3 or just v3? Maybe we could expect support before v3 release.

**Reply by adrianeboyd** [reply]
This looks great, thanks! We're only planning to add new languages/pipelines to v3 at this point, so we'll see if we can get everything integrated for the next rc release.

**Reply by adrianeboyd** [reply]
A few notes after looking at the data:

* we switched the built-in converters to be generator functions (in the upcoming 3.0.0rc3), so the `convert` memory usage is a bit lower (in my very rough test ~15G, which is not great, but better at least? in any case, splitting the original file into multiple conllu files would be the solution)
* the CoNLL-U files are missing `SpaceAfter=No`, which is what the converter uses to create the raw texts (I wrote a quick script to add it based on the texts)
* for the v3 project setup, leave out `--morphology` in `spacy convert` and use a `morphologizer` rather than a `tagger` in the pipeline

**Reply by kuk** [reply]
>  the convert memory usage is a bit lower (in my very rough test ~15G, which is not great, but better at least? in any case, splitting the original file into multiple conllu files would be the solution)

What is your RAM limit for training procedure? Also do you train on GPU or CPU? Does is bother you that training takes 2-3 hours per epoch on CPU?

1. We could split train dataset on our side and distribute it as tar.gz archive. But what is the optimal chunk size? What if you would need to change it in future? Maybe it would be better to have a script on your side that takes large .conllu file, chunk size as argument and generate splits?
2. Maybe it would be better just to make train dataset smaller. By shrinking dataset from 100k docs to 20k we loose 1 percentage points of accuracy. But training would take 5 times less RAM and time.
![image](https://user-images.githubusercontent.com/153776/104349551-1ed64d80-5514-11eb-8139-703c4b7da617.png)

> the CoNLL-U files are missing SpaceAfter=No, which is what the converter uses to create the raw texts (I wrote a quick script to add it based on the texts)

No problem, I can add SpaceAfter=No. Just curious, where do we need it? NER?

> for the v3 project setup, leave out --morphology in spacy convert and use a morphologizer rather than a tagger in the pipeline

Ok, will try

**Reply by adrianeboyd** [reply]
The `trf` models are trained on GPU, but all other models are trained on CPU. I think we currently train with 12G RAM by default, but we can switch to larger configurations as needed, which is also the case for the `xx` models. The `xx_ent_wiki` model has ~20M training tokens, which is similar in size to this data. The morphologizer+parser model will be trained separately from the NER model, so we'll just have to see what the memory usage looks like. We always set `max_steps` and it would be easy to truncate the training corpus in the config with the corpus reader if needed, so I'm sure we can get it working. To keep things simpler on our end (at least initially), we'll also use the same vectors config we use for our other pipelines, where `md` is pruned to 20K vectors and `lg` has all 500K vectors.

If we have raw texts, we can train and evaluate on actual rather than gold tokenization. It affects the training a little (the models ignore annotation when the spacy tokenization is misaligned with the gold tokenization), but the main reason is to be able to provide a more realistic evaluation since users are nearly always working from plain texts rather than gold tokenization. The v3 CoNLL-U converter also currently assumes that you are using `SpaceAfter=No`, so if it's missing it generates unrealistic raw texts with whitespace separation, which will throw off the tokenization and the evaluation.

It looks like the current Russian default tokenizer settings are okay to start with, with `token_f` at 0.978. The current Russian tokenizer isn't customized very much, so it might be easy to improve some easy cases around punctuation without much work. It might also be possible to write a custom `RussianTokenizer` that lets you switch between the spacy tokenizer and the same one used in the corpus, but it would be a little extra work to set up the options and the serialization, so I'll just go with the existing defaults for now.

**Reply by kuk** [reply]
I am planning
1. Split train into 10 chunks, each 10k docs. Replace train.conllu with train.conllu.tag.gz
2. Add SpaceAfter=No

Ok?

> To keep things simpler on our end (at least initially), we'll also use the same vectors config we use for our other pipelines, where md is pruned to 20K vectors and lg has all 500K vectors.

By pruning from 500k to 20k we loose ~1 percentage point of accuracy. Optimal threshold is around ~125k 
![image](https://user-images.githubusercontent.com/153776/104473359-f6128e80-55cd-11eb-87c3-654c0bc9c36e.png)

**Reply by buriy** [reply]
Hi @adrianeboyd , we indeed have a better rule-based tokenizer for spacy: https://github.com/aatimofeev/spacy_russian_tokenizer (2x less errors in average), but it quickly becomes slow when it has many rules, please see tokenization speed and quality comparison at https://github.com/natasha/naeval#tokenization .
Could we improve its speed while having even more rules?

Also, maybe you could share your typical training setup? (Or maybe it's published at some repository?)

**Reply by adrianeboyd** [reply]
The model size on disk is also a factor and for now, I'd like to keep the `md` models uniform.

~~Here's a preview of the `sm` model if you'd like to have a look: https://drive.google.com/file/d/1neef5lH_jmHJ_-s2q4wm4ww1ceEJsZt3/view?usp=sharing . Due to config changes, you'll need to install `develop` from source to load it.~~ [removed after official release]

The configs are the same for the `md` and `lg` models except for the vectors. We train the components separately as much as possible, so we have separate tasks for: tok2vec+tagger+morphologizer+parser, senter, and ner. The senter models took ~1 hour, the ner models ~2 hours, and the longest one (`md` morphologizer+parser) took ~8 hours, which is all fine.

**Reply by buriy** [reply]
@adrianeboyd Thanks a lot. I'm mostly interested in comparing to a configuration with all components set to work properly and then copying or finding the best hyper-params.

**Reply by kuk** [reply]
> I am planning
> Split train into 10 chunks, each 10k docs. Replace train.conllu with train.conllu.tag.gz
> Add SpaceAfter=No

Do I still need to add those features or training works fine on your side?

> The model size on disk is also a factor and for now, I'd like to keep the md models uniform

Okey

> Here's a preview of the sm model if you'd like to have a look

Looks good! Eval section in meta.json is even better then ours. How to access token morphology? Installed Spacy from development branch

```
>>> token.pos_, token.tag_

# spacy, spacy-nightly
NOUN, NOUN__Animacy=Anim|Case=Nom|Gender=Masc|Number=Sing
 
# development, ru_core_news_sm-3.0.0a1
NOUN, NOUN
```

**Reply by adrianeboyd** [reply]
We have everything set up on our end, so you don't need to modify anything for us. You might want to add `SpaceAfter=No` if you expect others to want to train their own spacy models from this data, though?

The morph information is under `token.morph`, see: https://nightly.spacy.io/api/morphology#morphanalysis

**Reply by kuk** [reply]
Ok, will keep dataset without changes for now. Currently you are the only consumer. Archived the repo https://github.com/natasha/natasha-spacy

**Comment by sdspieg** [comment]
This is GREAT news! Thanks to all involved indeed, and we can't wait to see it up there with the other languages...

**Comment by kuk** [comment]
Proposal got accepted in v3.0.0rc3 https://github.com/explosion/spaCy/releases/tag/v3.0.0rc3. See https://nightly.spacy.io/models/ru for official Russian model pretrained by Spacy team.
